x/gamm/twap/store_test.go

```diff
@@ -0,0 +1,181 @@
+// Then runs k.trackChangedPool on every item in the test case list.
+// Then checks that changed pools returns the list, deduplicated.
+//
+// This achieves testing the functionality that we depend on, that this clears every end block.
+func (s *TestSuite) TestTrackChangedPool() {
+	tests := map[string][]uint64{
+		"single":         {1},
+		"duplicated":     {1, 1},
+		"four":           {1, 2, 3, 4},
+		"many with dups": {1, 2, 3, 4, 3, 2, 1},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			// Test that no cumulative ID registers as tracked
+			changedPools := s.twapkeeper.GetChangedPools(s.Ctx)
+			s.Require().Empty(changedPools)
+
+			// Track every pool in list
+			cumulativeIds := map[uint64]bool{}
+			for _, v := range test {
+				cumulativeIds[v] = true
+				s.twapkeeper.TrackChangedPool(s.Ctx, v)
+			}
+
+			changedPools = s.twapkeeper.GetChangedPools(s.Ctx)
+			s.Require().Equal(len(cumulativeIds), len(changedPools))
+			for _, v := range changedPools {
+				s.Require().True(cumulativeIds[v])
+			}
+			s.Commit()
+		})
+	}
+}
+
+// TestGetAllMostRecentRecordsForPool takes a list of records as test cases,
+// and runs storeNewRecord for everything in sequence.
+// Then it runs GetAllMostRecentRecordsForPool, and sees if its equal to expected
+func (s *TestSuite) TestGetAllMostRecentRecordsForPool() {
+	tPlusOne := baseTime.Add(time.Second)
+	baseRecord := newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA")
+	tPlusOneRecord := newEmptyPriceRecord(1, tPlusOne, "tokenB", "tokenA")
+	tests := map[string]struct {
+		recordsToSet    []types.TwapRecord
+		poolId          uint64
+		expectedRecords []types.TwapRecord
+	}{
+		"set single record": {
+			[]types.TwapRecord{baseRecord},
+			1,
+			[]types.TwapRecord{baseRecord},
+		},
+		"query non-existent pool": {
+			[]types.TwapRecord{baseRecord},
+			2,
+			[]types.TwapRecord{},
+		},
+		"set two records": {
+			[]types.TwapRecord{baseRecord, tPlusOneRecord},
+			1,
+			[]types.TwapRecord{tPlusOneRecord},
+		},
+		"set two records, reverse order": {
+			// The last record, independent of time, takes precedence for most recent.
+			[]types.TwapRecord{tPlusOneRecord, baseRecord},
+			1,
+			[]types.TwapRecord{baseRecord},
+		},
+		"set multi-asset pool record": {
+			[]types.TwapRecord{
+				newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenB"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenA")},
+			1,
+			[]types.TwapRecord{
+				newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenB")},
+		},
+	}
+
+	for name, test := range tests {
+		s.Run(name, func() {
+			s.SetupTest()
+			for _, record := range test.recordsToSet {
+				s.twapkeeper.StoreNewRecord(s.Ctx, record)
+			}
+			actualRecords, err := s.twapkeeper.GetAllMostRecentRecordsForPool(s.Ctx, test.poolId)
```

Comment:
q: is `types` package a good location for `GetAllMostRecentTwapsForPool`? I think the only place where it is called right now is from keeper method that is unexported

Reply:
Agreed its a bit awkward there, but it has the advantage that all key formatting is encapsulated into that file.